File: app/src/androidTest/java/com/github/jameshnsears/quoteunquote/database/DatabaseTestHelper.java
package com.github.jameshnsears.quoteunquote.database;

import com.github.jameshnsears.quoteunquote.database.quotation.QuotationEntity;
import com.github.jameshnsears.quoteunquote.utils.ContentSelection;
import com.github.jameshnsears.quoteunquote.utils.widget.WidgetIdHelper;

import java.util.ArrayList;
import java.util.List;

public class DatabaseTestHelper {
    protected DatabaseRepositoryDouble databaseRepositoryDouble = DatabaseRepositoryDouble.getInstance();

    /*
        @Rule
    public InstantTaskExecutorRule instantTaskExecutorRule = new InstantTaskExecutorRule();

    @Rule
    public CountingTaskExecutorRule countingTaskExecutorRule = new CountingTaskExecutorRule();

    protected static int widgetID = 1;
    protected QuoteUnquoteModelFake quoteUnquoteModel;

    @Before
    public void setUp() {
        ApplicationProvider.getApplicationContext().deleteDatabase(AbstractHistoryDatabase.DATABASE_NAME);
        quoteUnquoteModel = new QuoteUnquoteModelFake();
    }

    @After
    public void tearDown() {
        quoteUnquoteModel.shutdown();
    }
     */

    public void insertTestDataSet01() {
        final List<QuotationEntity> quotationEntityList = new ArrayList<>();

        quotationEntityList.add(
                new QuotationEntity(
                        DatabaseRepository.DEFAULT_QUOTATION_DIGEST, "a0", "q0"
                ));
        quotationEntityList.add(
                new QuotationEntity(
                        "d1", "a1", "q1"
                ));

        databaseRepositoryDouble.insertQuotations(quotationEntityList);
    }

    public void insertTestDataSet02() {
        final List<QuotationEntity> quotationEntityList = new ArrayList<>();

        quotationEntityList.add(
                new QuotationEntity(
                        "d2", "a2", "q1"
                ));
        quotationEntityList.add(
                new QuotationEntity(
                        "d3", "a2", "q3"
                ));
        quotationEntityList.add(
                new QuotationEntity(
                        "d4", "a4", "q1"
                ));

        databaseRepositoryDouble.insertQuotations(quotationEntityList);
    }

    public void insertTestDataSet03() {
        final List<QuotationEntity> quotationEntityList = new ArrayList<>();

        quotationEntityList.add(
                new QuotationEntity(
                        "d5", "a5", "q1"
                ));
        quotationEntityList.add(
                new QuotationEntity(
                        "d6", "a2", "q6"
                ));

        databaseRepositoryDouble.insertQuotations(quotationEntityList);
    }

    public void setDefaultQuotation() {
        QuotationEntity defaultQuotation = databaseRepositoryDouble.getQuotation(DatabaseRepository.DEFAULT_QUOTATION_DIGEST);
        databaseRepositoryDouble.markAsPrevious(WidgetIdHelper.WIDGET_ID, ContentSelection.ALL, defaultQuotation.digest);
    }
}

File: app/src/androidTest/java/com/github/jameshnsears/quoteunquote/database/DatabaseTestHelper.kt
package com.github.jameshnsears.quoteunquote.database

import android.content.Context
import android.os.Build
import androidx.annotation.NonNull
import androidx.test.core.app.ApplicationProvider
import com.github.jameshnsears.quoteunquote.database.quotation.QuotationEntity
import com.github.jameshnsears.quoteunquote.utils.ContentSelection
import com.github.jameshnsears.quoteunquote.utils.preference.PreferencesFacade
import junit.framework.TestCase.assertTrue
import org.junit.Before

abstract class DatabaseTestHelper {
    @NonNull
    var databaseRepositoryDouble: DatabaseRepositoryDouble =
        DatabaseRepositoryDouble.getInstance(ApplicationProvider.getApplicationContext())

    @JvmField
    val context: Context = ApplicationProvider.getApplicationContext()

    @Before
    open fun before() {
        databaseRepositoryDouble.eraseAllDatabsaes()
        PreferencesFacade.erase(context)
        DatabaseRepository.useInternalDatabase = true
    }

    fun insertQuotationTestData01() {
        val quotationEntityList: MutableList<QuotationEntity> = ArrayList()
        quotationEntityList.add(
            QuotationEntity(
                DatabaseRepository.getDefaultQuotationDigest(),
                "w1",
                "a0",
                "q0",
            ),
        )
        quotationEntityList.add(QuotationEntity("d1234567", "w1", "a1", "q1"))
        databaseRepositoryDouble.insertQuotations(quotationEntityList)
    }

    fun insertQuotationTestData02() {
        val quotationEntityList: MutableList<QuotationEntity> = ArrayList()
        quotationEntityList.add(QuotationEntity("d2345678", "w1", "a2", "q1"))
        quotationEntityList.add(QuotationEntity("d3456789", "w1", "a2", "q3"))
        quotationEntityList.add(QuotationEntity("d4567890", "w1", "a4", "q1"))
        databaseRepositoryDouble.insertQuotations(quotationEntityList)
    }

    fun insertQuotationTestData03() {
        val quotationEntityList: MutableList<QuotationEntity> = ArrayList()
        quotationEntityList.add(QuotationEntity("d5678901", "w1", "a5", "q1"))
        quotationEntityList.add(QuotationEntity("d6789012", "w1", "a2", "q6"))
        databaseRepositoryDouble.insertQuotations(quotationEntityList)
    }

    fun setDefaultQuotationAll(widgetId: Int) {
        databaseRepositoryDouble.markAsPrevious(
            widgetId,
            ContentSelection.ALL,
            getDefaultQuotation().digest,
        )

        databaseRepositoryDouble.markAsCurrent(
            widgetId,
            getDefaultQuotation().digest,
        )
    }

    fun setDefaultQuotationAuthor(widgetId: Int) {
        databaseRepositoryDouble.markAsPrevious(
            widgetId,
            ContentSelection.AUTHOR,
            getDefaultQuotation().digest,
        )
    }

    fun setDefaultQuotationSearch(widgetId: Int) {
        databaseRepositoryDouble.markAsPrevious(
            widgetId,
            ContentSelection.SEARCH,
            getDefaultQuotation().digest,
        )
    }

    fun markDefaultQuotationAsFavourite() {
        databaseRepositoryDouble.markAsFavourite(getDefaultQuotation().digest)
    }

    fun getDefaultQuotation(): QuotationEntity {
        return databaseRepositoryDouble.getQuotation(DatabaseRepository.getDefaultQuotationDigest())
    }

    fun populateInternal(widgetId: Int) {
        DatabaseRepositoryDouble.useInternalDatabase = true

        insertInternalQuotations()

        databaseRepositoryDouble.markAsCurrent(
            widgetId,
            "d1234567",
        )

        databaseRepositoryDouble.markAsPrevious(
            widgetId,
            ContentSelection.ALL,
            "d1234567",
        )

        databaseRepositoryDouble.markAsFavourite(
            databaseRepositoryDouble.getCurrentQuotation(widgetId).digest,
        )

        assertTrue(databaseRepositoryDouble.countAll().blockingGet() == 5)
    }

    fun insertInternalQuotations() {
        DatabaseRepositoryDouble.useInternalDatabase = true

        insertQuotationTestData01()
        insertQuotationTestData02()
    }

    fun populateExternal(widgetId: Int) {
        DatabaseRepositoryDouble.useInternalDatabase = false

        insertExternalQuotations()

        databaseRepositoryDouble.markAsCurrent(
            widgetId,
            DatabaseRepositoryDouble.getDefaultQuotationDigest(),
        )

        databaseRepositoryDouble.markAsPrevious(
            widgetId,
            ContentSelection.ALL,
            DatabaseRepositoryDouble.getDefaultQuotationDigest(),
        )

        databaseRepositoryDouble.markAsFavourite(
            databaseRepositoryDouble.getCurrentQuotation(widgetId).digest,
        )

        assertTrue(databaseRepositoryDouble.countAll().blockingGet() == 2)
    }

    fun insertExternalQuotations() {
        DatabaseRepositoryDouble.useInternalDatabase = false

        val quotationEntityList: MutableList<QuotationEntity> = ArrayList()
        quotationEntityList.add(
            QuotationEntity(
                DatabaseRepository.getDefaultQuotationDigest(),
                "",
                "external_a0",
                "external_q0",
            ),
        )
        quotationEntityList.add(
            QuotationEntity(
                "00000001",
                "",
                "external_a1",
                "external_q1",
            ),
        )
        databaseRepositoryDouble.insertQuotations(quotationEntityList)
    }

    protected fun canWorkWithMockk() = Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q
}
